### PR TITLE
Trev8939/download preplanned map area

### PR DESCRIPTION
**Pull request by TADraeseke**
Kotlin to java conversion of the download preplanned map area sample. There's a fair bit of code to go through here when you can. Thanks! @hud10837!

### CODE REVIEW

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+                }
+                downloadPreplannedOfflineMapResult.tableErrors?.forEach { (key, value) ->
+                  stringBuilder.append("Table: ").append(key.tableName).append(". Exception: ")
+                    .append(value.message).append(". ")
+                }
+                val error =
+                  "One or more errors occurred with the Offline Map Result: $stringBuilder"
+                Toast.makeText(this, error, Toast.LENGTH_LONG).show()
+                Log.e(TAG, error)
+                return@addJobDoneListener
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:184]: If it's necessary I understand, but all of the append statements make this quite difficult to understand. I suggest this as a simplification, but I'm not 100% on how stringbuilder operates.

```suggestion
                val stringBuilder = StringBuilder("Errors: ")
                downloadPreplannedOfflineMapResult.layerErrors?.forEach { (key, value) ->
                  stringBuilder.append("Layer: ${key.name}. Exception: ${value.message}. ")
                }
                downloadPreplannedOfflineMapResult.tableErrors?.forEach { (key, value) ->
                  stringBuilder.append("Table: ${key.tableName}. Exception: ${value.message}. ")
                }
                val error =
                  "One or more errors occurred with the Offline Map Result: $stringBuilder"
                Toast.makeText(this, error, Toast.LENGTH_LONG).show()
                Log.e(TAG, error)
                return@addJobDoneListener
```

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+            if (preplannedMapArea.loadStatus == LoadStatus.LOADED) {
+              // add the area of interest as a graphic
+              mapView.graphicsOverlays[0].graphics.add(Graphic(preplannedMapArea.areaOfInterest))
+            } else {
+              val error =
+                "Failed to load preplanned map area: " + preplannedMapArea.loadError.message
+              Toast.makeText(this, error, Toast.LENGTH_LONG).show()
+              Log.e(TAG, error)
+            }
+          }
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:248]: Could these `forEach`s perhaps be combined?

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        }
+        // on list view click
+        availableAreasListView.onItemClickListener =
+          AdapterView.OnItemClickListener { _: AdapterView<*>?, _: View?, i: Int, _: Long ->
+            mSelectedPreplannedMapArea = preplannedMapAreas[i]
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:253]: Hungarian notation :)

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+            if (mSelectedPreplannedMapArea != null) {
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:254]: I would early exit here again

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+                    + File.separator + mSelectedPreplannedMapArea?.portalItem?.title
+              ).let {
+                downloadButton.isEnabled = !it.exists()
+              }
+            } else {
+              downloadButton.isEnabled = false
+            }
+          }
+      } catch (e: Exception) {
+        val error = "Failed to get the Preplanned Map Areas from the Offline Map Task."
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:280]:
```suggestion
        val error = "Failed to get the preplanned map areas from the offline map task."
```

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+  private var downloadPreplannedOfflineMapJob: DownloadPreplannedOfflineMapJob? = null
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // delete any previous instances of downloaded maps
+    externalCacheDir?.deleteRecursively()
+
+    // create up a temporary directory in the app's cache when required
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:74]:
```suggestion
    // create a temporary directory in the app's cache when required
```

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+    // set the authentication manager to handle challenges when accessing the portal
+    // Note: The sample data is publicly available, so you shouldn't be challenged
+    AuthenticationManager.setAuthenticationChallengeHandler(
+      DefaultAuthenticationChallengeHandler(this)
+    )
+
+    // create a portal to ArcGIS Online
+    val portal = Portal(getString(R.string.arcgis_online_url))
+    // create a portal item using the portal and the item id of a map service
+    val portalItem = PortalItem(portal, getString(R.string.naperville_water_network_url))
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:92]: Perhaps its best to rename this string resource to something like `naperville_water_network_id`

File: kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+    // create download button
+    downloadButton.apply {
+      isEnabled = false
+      setOnClickListener { downloadPreplannedArea(offlineMapTask) }
+    }
+  }
+
+  /**
+   * Download the selected preplanned map area from the list view to a temporary directory. The
+   * download job is tracked in another list view.
```
Comment (hud10837) [inline kotlin/download-preplanned-map-area/src/main/java/com/esri/arcgisruntime/sample/downloadpreplannedmaparea/MainActivity.kt:126]: I think you still need KDoc @params here

**Comment by TADraeseke** [comment]
Thanks @hud10837!

**Pull request merged**